x/gmp/module_ibc.go

```diff
@@ -0,0 +1,205 @@
+}
+
+// OnChanOpenConfirm implements the IBCModule interface
+func (im IBCModule) OnChanOpenConfirm(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+) error {
+	return im.app.OnChanOpenConfirm(ctx, portID, channelID)
+}
+
+// OnChanCloseInit implements the IBCModule interface
+func (im IBCModule) OnChanCloseInit(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+) error {
+	return im.app.OnChanCloseInit(ctx, portID, channelID)
+}
+
+// OnChanCloseConfirm implements the IBCModule interface
+func (im IBCModule) OnChanCloseConfirm(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+) error {
+	return im.app.OnChanCloseConfirm(ctx, portID, channelID)
+}
+
+// OnRecvPacket implements the IBCModule interface
+func (im IBCModule) OnRecvPacket(
+	ctx sdk.Context,
+	modulePacket channeltypes.Packet,
+	relayer sdk.AccAddress,
+) ibcexported.Acknowledgement {
+	// this line is used by starport scaffolding # oracle/packet/module/recv
+
+	var data transfertypes.FungibleTokenPacketData
+	var err error
+	if err := types.ModuleCdc.UnmarshalJSON(modulePacket.GetData(), &data); err != nil {
+		ctx.Logger().Debug(fmt.Sprintf("cannot unmarshal ICS-20 transfer packet data: %s", err.Error()))
+		return im.app.OnRecvPacket(ctx, modulePacket, relayer)
+	}
+
+	var msg Message
+	if err = json.Unmarshal([]byte(data.GetMemo()), &msg); err != nil {
+		ctx.Logger().Debug(fmt.Sprintf("cannot unmarshal memo: %s", err.Error()))
+		return im.app.OnRecvPacket(ctx, modulePacket, relayer)
+	}
+
+	if msg.Payload == nil {
+		return im.app.OnRecvPacket(ctx, modulePacket, relayer)
+	}
+
+	switch msg.Type {
+	case TypeGeneralMessage:
+		ctx.Logger().Info(fmt.Sprintf("Got pure general message: %v", msg))
+		return im.app.OnRecvPacket(ctx, modulePacket, relayer)
+	case TypeGeneralMessageWithToken:
+		ctx.Logger().Info(fmt.Sprintf("Got general message with token: %v", msg))
+		payloadType, err := abi.NewType("string", "", nil)
+		if err != nil {
+			ctx.Logger().Info(fmt.Sprintf("failed to create reflection: %s", err.Error()))
+			return channeltypes.NewErrorAcknowledgement(cosmossdkerrors.Wrapf(transfertypes.ErrInvalidMemo, "unable to define new abi type (%s)", err.Error()))
+		}
+
+		// Add 4 bytes to the payload to match the length of the payload otherwise the unpack will fail
+		payloadData := msg.Payload
+		ctx.Logger().Info(fmt.Sprintf("ABI-encoded data: %v", hex.EncodeToString(payloadData)))
+
+		payloadData = append(make([]byte, 4), payloadData...)
+		ctx.Logger().Info(fmt.Sprintf("Padded ABI-encoded data: %v", hex.EncodeToString(payloadData)))
+
+		args, err := abi.Arguments{{Type: payloadType}}.Unpack(payloadData)
+		if err != nil {
+			ctx.Logger().Info(fmt.Sprintf("failed to unpack: %s", err.Error()))
+			return channeltypes.NewErrorAcknowledgement(cosmossdkerrors.Wrapf(transfertypes.ErrInvalidMemo, "unable to unpack payload (%s)", err.Error()))
+		}
+		ctx.Logger().Info(fmt.Sprintf("Unpacked data: %+v", args))
```

Comment:
It would probably be better to have this kind of debugging messages as a lower loglevel, not info.